### PR TITLE
feature: calculate subvolume defocus values

### PR DESCRIPTION
**Pull request by McHaillet**
This PR adds support for calculating defocus value for each subvolume specified with `--volume-split` in template matching. As particles experience a different defocus based on their position in the ice and depending on the tilt angle, this potentially provides detailed defocus gradient correction. However, I also added a page on the site with a benchmark for measuring the effects of defocus gradient correction and found that the effect is very small. I thought this would be cool to add.

- [x] implementation
- [x] add benchmark 

* I left the version number as we did not release 0.7.2 yet.
* I made a change in angles.py to read the log_level as an integer as this is what is available in the TMJob. I also removed a section there that overwrote the log_level as a string. (should have put this in a separate PR, sorry for that)

### CODE REVIEW

File: src/pytom_tm/angles.py [hunk 1]
```diff
@@ -113,7 +113,7 @@ def get_angle_list(
         then X, then Z2
     symmetry: int, default 1
         the returned list will only have Z2 angles [0, (2*pi/symmetry))
-    log_level: str, default 'DEBUG'
+    log_level: int, default 'DEBUG'
```
Comment (McHaillet) [inline src/pytom_tm/angles.py:116]:
```suggestion
    log_level: int, default logging.DEBUG
```

File: src/pytom_tm/tmjob.py [hunk 2]
```diff
@@ -92,6 +93,64 @@ def load_json_to_tmjob(
...
+        [
+            patch_center_z,
+        ]
+        * n_tilts
+    )
+    ta_array = np.array(tilt_angles)
+    if angles_in_degrees:
+        ta_array = np.deg2rad(ta_array)
+    if invert_handedness:
+        ta_array *= -1
```
Comment (sroet) [inline src/pytom_tm/tmjob.py:149]: not tested, please add a test

File: src/pytom_tm/tmjob.py [hunk 3]
```diff
@@ -92,6 +93,64 @@ def load_json_to_tmjob(
...
+            patch_center_x,
+        ]
+        * n_tilts
+    )
+    z_centers = np.array(
+        [
+            patch_center_z,
+        ]
+        * n_tilts
+    )
```
Comment (sroet) [inline src/pytom_tm/tmjob.py:144]: This feels surprisingly wastefull in lines, also probably quicker (and cleaner in my opinion) to build with [`np.full`](https://numpy.org/doc/stable/reference/generated/numpy.full.html)
```suggestion
    x_centers = np.full(n_tilts, patch_center_x)
    z_centers = np.full(n_tilts, patch_center_z)
```

File: src/pytom_tm/tmjob.py [hunk 4]
```diff
@@ -809,8 +869,33 @@ def start_job(
...
         if self.tilt_angles is not None:
-            # for the tomogram a binary wedge is generated to explicitly set the
-            # missing wedge region to 0
+            if self.tilt_weighting and self.defocus_handedness != 0:
+                # adjust ctf parameters for this specific patch in the tomogram
+                full_tomo_center = np.array(self.tomo_shape) / 2
+                patch_center = (
+                    np.array(self.search_origin) + np.array(self.search_size) / 2
+                )
+                relative_patch_center_um = (patch_center - full_tomo_center) * 1e-4
```
Comment (sroet) [inline src/pytom_tm/tmjob.py:878]: isn't this pixel size dependent?

File: src/pytom_tm/tmjob.py [hunk 5]
```diff
@@ -809,8 +869,33 @@ def start_job(
...
+                    relative_patch_center_um[0],
+                    relative_patch_center_um[2],
+                    self.tilt_angles,
+                    angles_in_degrees=True,
+                    invert_handedness=self.defocus_handedness < 0,
+                )
+                for ctf, defocus_shift in zip(self.ctf_data, defocus_offsets):
+                    ctf["defocus"] = ctf["defocus"] + defocus_shift * 1e-6
+                logging.debug(
+                    f"Patch center (nr. of voxels): "
```
Comment (sroet) [inline src/pytom_tm/tmjob.py:889]: no an fstring needed
```suggestion
                    "Patch center (nr. of voxels): "
```

File: src/pytom_tm/tmjob.py [hunk 6]
```diff
@@ -809,8 +869,33 @@ def start_job(
...
+                    f"{np.array_str(relative_patch_center_um * 1e4, precision=2)}"
+                )
+                logging.debug(
+                    f"Defocus values (um): "
```
Comment (sroet) [inline src/pytom_tm/tmjob.py:893]: not fstring needed
```suggestion
                    "Defocus values (um): "
```

File: src/pytom_tm/tmjob.py [hunk 7]
```diff
@@ -809,8 +869,33 @@ def start_job(
 
         # if tilt angles are provided we can create wedge filters
         if self.tilt_angles is not None:
-            # for the tomogram a binary wedge is generated to explicitly set the
-            # missing wedge region to 0
+            if self.tilt_weighting and self.defocus_handedness != 0:
```
Comment (sroet) [inline src/pytom_tm/tmjob.py:872]: This whole if statement is untested, please add tests

**Review by sroet — CHANGES_REQUESTED** [review]
Some issues and feedback.

Other points:
- Would this need an option to override the tilt-axis? Or are those always "close enough" to the y-axis?

File: docs/benchmarks/defocus_handedness.md [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+the defocus 
+handedness of a tomogram.
+
+## Approach
+
+<figure markdown="span">
+  ![annotations](defocus_handedness_figures/blik_view.png){ width="400" }
+  <figcaption>Initial view of the data: annotations (turqoise sphere) made by 
+pytom-match-pick on tomogram 27 of EMPIAR-10985.</figcaption>
+</figure>
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply docs/benchmarks/defocus_handedness.md:31]: I guess it is not supported in native markdown: https://squidfunk.github.io/mkdocs-material/reference/images/#image-captions

The md_in_html extension (in the mkdocs.yml) enables it.

File: docs/benchmarks/defocus_handedness.md [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+### Implementation of defocus gradient
+
+An implementation for the defocus gradient calculation is available in our 
+repository in `src/pytom_tm/TMJob.py`, see function 
+`get_defocus_offsets()`. 
+
+### Tilt series preprocessing
+
+Download tilt-series 27 from EMPIAR-10985:
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply docs/benchmarks/defocus_handedness.md:154]:
```suggestion
Download tilt-series 27 from [EMPIAR-10985](https://doi.org/10.6019/EMPIAR-10985):
```

File: docs/benchmarks/defocus_handedness.md [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+Fit ctf with IMOD; in the GUI select `fit each view separately` and then `autofit all 
+views`. Tilt-axis angle is set to '-3.2323' as optimized by AreTomo (check the .aln file).
+
+```commandline
+ctfplotter -input raw_data/9x9_ts_27_sort.mrc -angleFn metadata/9x9_ts_27_sort.rawtlt -aAngle -3.2323 -defFn metadata/9x9_ts_27_sort.defocus -pixelSize 0.107 -volt 300 -cs 2.7  -expDef 3000 -range -60,60
+```
+
+### Creating a template and mask
+
+I downloaded the subtomogram average (EMD-33115) calculated for the EMPIAR dataset. NOTE: The reference is mirrored as I noticed in the first run without mirroring that the scores were unexpectedly poor. Mirroring fixes this issue.
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply docs/benchmarks/defocus_handedness.md:201]:
```suggestion
I downloaded the subtomogram average [EMD-33115](https://www.ebi.ac.uk/emdb/EMD-33115) calculated for the EMPIAR dataset. NOTE: The reference is mirrored as I noticed in the first run without mirroring that the scores were unexpectedly poor. Mirroring fixes this issue.
```

File: docs/benchmarks/defocus_handedness.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (McHaillet) [reply docs/benchmarks/defocus_handedness.md:31]: I meant the captions are not supported

File: src/pytom_tm/tmjob.py [hunk 11]
```diff
@@ -419,10 +484,6 @@ def __init__(
             else:
                 angle_increment = 7.0
         self.rotation_file = angle_increment
-        if job_loaded_for_extraction:
-            log_level = "DEBUG"
-        else:
-            log_level = "INFO"
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/tmjob.py:425]: You are right! I added a comment in the code to explain. The logic was fine but it was missing with log_level parameter that was passed to the tmjob. I think its fixed now and hopefully clear from the comment for future development.

File: src/pytom_tm/tmjob.py (hunk 4 above)
Comment (McHaillet) [reply src/pytom_tm/tmjob.py:872]: Thanks for spotting! you are absolutely right, completely overlooked that.

Comment (McHaillet) [reply src/pytom_tm/tmjob.py:872]: I will need to recalculate the results

**Comment by McHaillet** [comment]
>     * Would this need an option to override the tilt-axis? Or are those always "close enough" to the y-axis?

The tilt axis of the tilt series is always aligned with the y-axis of the tomogram. It's the common convention as far as I know. So should be safe.

**Review by sroet — APPROVED** [review]
Thanks for the changes! LGTM!

**Pull request merged**